application/bsdd_utils.py

```diff
@@ -42,4 +50,64 @@ def get_inst(instance_id):
     
     return hierarchical_bsdd_results     
 
+def get_processed_bsdd_table(bsdd_results, session, schema):
+    """
+    The function returns data for the 'bsdd data' table in a file metrics report. 
+    More specifically,it returns a list of dictionaries containing the classifications (ifc instances, in this case) 
+    and their valid/invalid counts.
+    """
+    bsdd_instances = [bsdd_table(result, session, schema) for result in bsdd_results]
+    bsdd_data = defaultdict(lambda: {'valid': 0, 'invalid': 0, 'domain_source': 'classification not found'})
+
+    for item in bsdd_instances:
+        classification = item['classification']
+        validity = item['validity']
+        bsdd_data[classification][validity] += 1
+
+    bsdd_data = [{**{'classification': k}, **v} for k, v in bsdd_data.items()]
+    return bsdd_data
+
+def instance_supertypes(observed_type, schema):
+    allowed_types = [observed_type]
+
+    while True:
+        try:
+            result = (lambda x: ifcopenshell.ifcopenshell_wrapper.schema_by_name(schema).declaration_by_name(x).supertype().name())(allowed_types[-1])
```

Comment:
Not sure I understand what this line does.
Also, no need to recompute the validity with respect to the type (nor any other criteria) it's available by doing `bsdd_result.val_ifc_type`